Run existing agent tests against the rust-keylime agent

mbestavros opened the pull request:
The current `test_restful.py` test suite runs a number of tests against the Python agent. This pull request adds a number of helper functions and a new set of tests that:

- shut down the Python agent
- start up the Rust agent
- run agent-specific tests against the Rust agent

Currently, one of the tests (`agent_quotes_identity_get`) passes, while others fail (mostly due to missing APIs in `rust-keylime`). I've added skips to the broken tests with documentation on the relevant `rust-keylime` issues blocking them, so it should still pass without fail.

Signed-off-by: Mark Bestavros <mbestavr@redhat.com>

aplanas commented:
This will block the Python code changes on issues on the rust side.

mpeters commented:
> This will block the Python code changes on issues on the rust side.

This is true and one of the reasons this won't be merged as-is currently. But eventually the rust-agent will become the official keylime agent and should block changes to the python side that break the rust side.

mbestavros commented:
@mpeters I'm marking this as ready. I've tidied everything up so there should be minimal code duplication, made sure it passes lint, added skips (with notes and links) to the broken tests, and fixed a bug with Rust agent invocation on the Github runners. Looks like all tests pass now.

aplanas commented on test/run_tests.sh:
```diff
@@ -163,6 +163,8 @@ echo "==========================================================================
 echo $'\t\t\tInstalling test requirements'
 echo "=================================================================================="
 pip3 install $UMODE_OPT -r $KEYLIME_DIR/test/test-requirements.txt
+git clone https://github.com/keylime/rust-keylime.git $KEYLIME_DIR/../rust-keylime
```
Can we use a variable, that if is set will do the clone and the cargo build? Something line `RUST_TEST` set as 0 in the script but overwritten from the value in the environment, as `RUST_TEST="${RUST_TEST:-0}"`

This way by default we do not clone and call `cargo build` when running the Python tests.  There is no guarantee that the environment will even have Rust installed when building and testing a Python program.

mbestavros replied:
Very true - I've added this in my latest push, and gated the Rust tests and build behind it.

aplanas commented on test/test_restful.py:
```diff
@@ -276,6 +276,33 @@ def initthread():
     return True
 
 
+def launch_rust_cloudagent():
```
Can we do a `@unittest.skipIf`? I do no think that is OK for a Python test to unconditionally run a test that depends on cargo and the Rust toolchain.

mbestavros replied:
Done! See above.

aplanas commented on test/test_restful.py:
```diff
@@ -276,6 +276,33 @@ def initthread():
+    """Start up the cloud agent"""
+    global agent_process, script_env, FORK_ARGS
+    if agent_process is None:
+        agent_process = subprocess.Popen(["cargo", "run", "--manifest-path", f"{PACKAGE_ROOT}/../rust-keylime/Cargo.toml", "--bin", "keylime_agent"],
```
Why do not call directly `keylime_agent`, instead of running it from `cargo run`?

mbestavros replied:
Good point - I just didn't know how to call it directly, but it's in the `.cargo/bin` directory. Added.

aplanas commented on test/test_restful.py:
```diff
@@ -915,6 +945,42 @@ def test_060_cv_version_get(self):
         self.assertEqual(results['supported_versions'], api_version.all_versions())
 
+    # Rust agent testset
+    def test_070_rust_agent_setup(self):
```
`skipIf`?

mbestavros replied:
Added!

aplanas commented on test/test_restful.py:
```diff
@@ -276,6 +277,33 @@ def initthread():
+    """Start up the cloud agent"""
+    global agent_process, script_env, FORK_ARGS
+    if agent_process is None:
+        agent_process = subprocess.Popen("/.cargo/bin/keylime_agent",
```
Do you think that the rust agent will be in the root? Is because is inside a container?

mbestavros replied:
I realized I pushed a broken path, which I've now fixed. It should be correct now.

aplanas commented on test/test_restful.py:
```diff
@@ -276,6 +277,33 @@ def initthread():
+                sys.stdout.write('\n\033[94m' + line + '\033[0m')
+        t = threading.Thread(target=initthread)
+        t.start()
+        time.sleep(10)
```
I guess that this sleep is here to wait for the agent to do the bind. This can produce flaky tests (tests that sometimes fails).

Maybe a better option is to wait (not tested):

```python
for retry in range(10):
    r = subprocess.run(["ss", "-l", "-t", "-n", "-p",  "( sport = :9002 )"], stdout=subprocess.PIPE)
    if b':9002" in r.stdout:
        break
    time.sleep(1)
else:
   raise Exception("Rust keylime_agent failed to run")
```

mbestavros replied:
This helper was copy-pasted from the `launch_cloudagent` function, which also does it this way. Likewise for the 10-second wait after initialization.

I'm not opposed to changing it, but it would also make sense to do it for the Python launch code too.

aplanas commented on test/test_restful.py:
```diff
@@ -915,6 +946,49 @@ def test_060_cv_version_get(self):
+        # Kill the Python agent and launch the Rust agent!
+        kill_cloudagent()
+        launch_rust_cloudagent()
+        time.sleep(10)
```
this second sleep can be skipped, if you use the "wait" loop. In any case there were a sleep inside the launch command.

mbestavros replied:
See above

aplanas commented on test/test_restful.py:
```diff
@@ -915,6 +946,49 @@ def test_060_cv_version_get(self):
+
+        # Kill the Python agent and launch the Rust agent!
+        kill_cloudagent()
+        launch_rust_cloudagent()
```
you launch the agent as a side effect of this test, you can consider to put it in `setUp`, and the kill in `tearDown`

mbestavros replied:
Putting the invocation in `setUp` and `tearDown` means we'd have to restart the agent every single test run, which would be extremely slow. Plus, some tests require the agent to be running between them, so that wouldn't work anyway.

It _might_ make sense to do it in `setUpClass`, but since we have to run separate versions of the agent at different times for different tests, that doesn't work at the moment either.

It's in a separate test so we know it's invoked before the rest of the Rust test suite. This is necessary because one of the reused tests (020) includes its own call to `launch_cloudagent`. If the Rust agent isn't running when that call happens, the Python agent will be booted up instead.

mpeters replied:
Even if we can't put it in `setUp()` and `tearDown()` it might make sense to pull it out into it's own function that checks to see if the agent is running and if not, starts the agent (if it's not running, and does a loop to make sure it's up before returning). Then this function can be included in every test that needs that agent. This would allow developers to be able to run single test methods as they are developing (a common pattern in most projects).

Then we could handle stopping the agents in the class teardown if they are running.

I know this is a change to the way things were for the other agent (the existing tests were fragile in this way), but it's becoming more necessary to fix if we copy that fragility into more places.

mbestavros replied:
@mpeters Okay, I'll give that a try. I think the cleanest way might actually just be to improve the existing `launch_cloudagent` function - it already checks if there's an agent running. If we provide a parameter to toggle between launching the Python and Rust agents, and incorporate @aplanas' suggestion around waiting until it's up, we could just include it in front of all tests that need an agent, as you say. Then, for the Rust tests, we just have one extra that stops it and restarts the Rust agent - that should default to Python for now and allow Rust when needed.

mpeters commented on test/test_restful.py:
```diff
@@ -100,6 +100,7 @@ def cmp(a, b):
 reg_process = None
 agent_process = None
 tenant_templ = None
+RUST_TEST = not bool(os.getenv("RUST_TEST"))
```
Why is this `not`? If you're negating the env var value shouldn't this var be called something like `SKIP_RUST_TEST`?

mpeters commented on test/test_restful.py:
```diff
@@ -275,7 +282,14 @@ def initthread():
+            elif retry == 10:
+                raise Exception(f"{agent} keylime_agent failed to launch")
+            time.sleep(1)
+        #time.sleep(10)
```
You can just remove this commented out code. It's safe in the git history.

mbestavros replied:
Done!

mpeters reviewed (changes requested):
Looking good. Some small changes/clarifications requested.

mbestavros commented:
Current status is that some cross-agent tests fail with this error message:

```
Error in test.test_restful.TestRestful.test_071_agent_keys_pubkey_get
  File "/usr/lib/python3.9/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3.9/site-packages/urllib3/connectionpool.py", line 381, in _make_request
    self._validate_conn(conn)
  File "/usr/lib/python3.9/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
    conn.connect()
  File "/usr/lib/python3.9/site-packages/urllib3/connection.py", line 362, in connect
    self.sock = ssl_wrap_socket(
  File "/usr/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 384, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib64/python3.9/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/lib64/python3.9/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/usr/lib64/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1129)
```

I've talked about this offline with @lkatalin, and she suggests that it may be a bug introduced by `rust-keylime` introducing mTLS, and might be fixed by this upstream PR: https://github.com/keylime/rust-keylime/pull/316

In any case, it seems like the test failure is down to upstream bugs and not due to these test changes - and, in any case, the Rust tests will not be run as part of the test suite yet.

mbestavros commented:
A quick summary of the changes I've made:

- Rewrote `launch_cloudagent()` as detailed [here](https://github.com/keylime/keylime/pull/812#discussion_r793987819); now, it will invoke the agent if it isn't already running, and can start either the Python or Rust agent. It also is a lot smarter about waiting for the agent to boot (instead of waiting 10 seconds, it checks if the agent's socket is open). This means we don't need a separate `launch_rust_cloudagent()` function anymore, and individual tests should be a bit more stable.
- Renamed `RUST_TEST` to `SKIP_RUST_TEST` in `test_restful` to better reflect what it does
- Added `launch_cloudagent()` invocations in front of all the Rust tests
- Addressed various suggestions/fixes from @mpeters

aplanas commented:
@mbestavros very cool changes! Thanks a lot!

mbestavros commented:
@aplanas Thanks for your reviews! They were very helpful 😄
